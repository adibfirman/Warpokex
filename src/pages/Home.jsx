import React, { useState } from 'react'
import { GET_ALL_POKEMONS, GET_POKEMON } from 'services/Query'
import client from 'services/graph'
import { useQuery } from '@apollo/react-hooks'
import { ListPokemon, Search } from 'components'
import { Button } from 'atomize'
/**
 * @todo
 * [need identifier] useApolloClient not work
 * if fixed remove client from 'services/graph'
 */
export default function Home(props) {
  const [query, setQuery] = useState(GET_ALL_POKEMONS)
  const [variables, setVariables] = useState({ first: 10, name: '' })
  const [search, setSearch] = useState('')

  const { loading, error, data, fetchMore } = useQuery(query, {
    client,
    variables
  })

  const handleItemClick = (data, e) => {
    props.history && props.history.push(`/pokemon/${data.name}`)
  }

  const handleChange = e => {
    e.persist()
    setSearch(e.target.value)
  }

  const handleEnter = e => {
    if (e.key === 'Enter') {
      setVariables({ ...variables, name: search })
      setQuery(
        ['', undefined, null].includes(search) ? GET_ALL_POKEMONS : GET_POKEMON
      )
    }
  }

  // console.log(data)
  const { pokemons, pokemon } = data

  const handleFetchMore = variables => {
    // console.log(variables)
    fetchMore({
      variables,
      updateQuery: (prev, { fetchMoreResult, variables }) => {
        if (!fetchMoreResult) return prev
        setVariables(variables)
        return Object.assign({}, prev, {
          pokemons: [...fetchMoreResult.pokemons],
          pokemon: [fetchMoreResult.pokemon]
        })
      }
    })
  }

  if (error) return 'Something Wrong'

  // console.log('current data', data)
  return (
    <>
      <Search
        isLoading={loading}
        onKeyPress={handleEnter}
        onChange={handleChange}
      />
      <ListPokemon
        onClick={handleItemClick}
        data={
          pokemons
            ? pokemons
            : pokemon
            ? Array.isArray(pokemon)
              ? pokemon
              : [pokemon]
            : []
        }
      />
      <Button
        data-testid="loadMore"
        onClick={e => {
          handleFetchMore({
            ...variables,
            first: variables.first + 5
          })
        }}
        disabled={loading}
        isLoading={loading}
        w="100%"
        bg="warning700"
        hoverBg="warning800"
        rounded="circle"
        color={'white'}
        p={{ l: '3rem', r: '2rem' }}
        m={{ b: '2rem' }}
      >
        {loading ? 'Loading...' : 'Load More'}
      </Button>
    </>
  )
}

Home.defaultProps = {
  query: GET_ALL_POKEMONS
}
