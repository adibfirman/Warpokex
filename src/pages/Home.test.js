/* global render wait */
import React from 'react'
import { MockedProvider } from 'react-apollo/test-utils';
import Home from 'pages/Home'
import mocks from 'pages/Pokemon.fixtures'

it('render Home page with the data', async () => {
    const {container, queryAllByTestId, getByTestId} = render(
        <MockedProvider mocks={mocks} addTypeName={false}>
            <Home/>
        </MockedProvider>
    )

    const listPokemon = getByTestId('ListPokemon')
    expect(listPokemon).toBeEmpty()

    // wait response 
    await wait(() => {
        expect(listPokemon).not.toBeEmpty() 
        expect(listPokemon.childElementCount).toBe(mocks[0].result.data.pokemons.length)
        const pokemonImage = queryAllByTestId('ListPokemon.image')
        expect(pokemonImage).toHaveLength(mocks[0].result.data.pokemons.length)
    })

    expect(container).toMatchSnapshot()
})
