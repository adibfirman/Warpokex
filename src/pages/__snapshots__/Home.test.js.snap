// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Home page with the data 1`] = `
<div>
  <div
    class="sc-bdVaJa yCdlf"
    data-testid="ListPokemon"
  >
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/bulbasaur.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          001
           | 
          Bulbasaur
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa ksqsBw"
          >
            Grass
          </span>
          <span
            class="sc-bdVaJa fNilym"
          >
            Poison
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/ivysaur.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          002
           | 
          Ivysaur
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa ksqsBw"
          >
            Grass
          </span>
          <span
            class="sc-bdVaJa fNilym"
          >
            Poison
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/venusaur.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          003
           | 
          Venusaur
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa ksqsBw"
          >
            Grass
          </span>
          <span
            class="sc-bdVaJa fNilym"
          >
            Poison
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/charmander.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          004
           | 
          Charmander
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa ijVPke"
          >
            Fire
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/charmeleon.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          005
           | 
          Charmeleon
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa ijVPke"
          >
            Fire
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/charizard.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          006
           | 
          Charizard
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa ijVPke"
          >
            Fire
          </span>
          <span
            class="sc-bdVaJa bHsrMp"
          >
            Flying
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/squirtle.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          007
           | 
          Squirtle
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa LEVCA"
          >
            Water
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/wartortle.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          008
           | 
          Wartortle
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa LEVCA"
          >
            Water
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/blastoise.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          009
           | 
          Blastoise
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa LEVCA"
          >
            Water
          </span>
        </label>
      </div>
    </div>
    <div
      class="sc-EHOje gxRZMW sc-bdVaJa geNTph"
    >
      <div
        class="sc-bdVaJa epdytx"
      >
        <img
          class="sc-gzVnrw hWZxqh"
          data-testid="ListPokemon.image"
          src="https://img.pokemondb.net/artwork/caterpie.jpg"
        />
        <h1
          class="sc-iwsKbI kCjaUS"
          color="black"
          data-testid="ListPokemon.title"
        >
          010
           | 
          Caterpie
        </h1>
        <label
          class="sc-bdVaJa cmuZrv"
          color="black"
        >
          <span
            class="sc-bdVaJa fNilym"
          >
            Bug
          </span>
        </label>
      </div>
    </div>
  </div>
  <button
    bg="warning700"
    class="sc-htpNat ccLvux"
    color="white"
    data-testid="loadMore"
  >
    Load More
  </button>
</div>
`;
